### ISSUE TITLE
`bugprone-unintended-char-ostream-output` reported with sufficient cast

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#133425]
```cpp
#include <cstdint>
#include <iostream>

enum E : std::uint8_t {
    E1
};

void f(E e)
{
    std::cout << static_cast<unsigned char>(e);
}
```

```
<source>:10:15: warning: 'unsigned char' passed to 'operator<<' outputs as character instead of integer. cast to 'unsigned int' to print numeric value or cast to 'char' to print as character [bugprone-unintended-char-ostream-output]
   10 |     std::cout << static_cast<unsigned char>(e);
      |               ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                  static_cast<unsigned int>(static_cast<unsigned char>(e))
```
https://godbolt.org/z/vzPoa9jnd

CC @HerrCai0907

**Comment by HerrCai0907** [comment]
I think it should be warned by default. since in this case, output are still invalid ascii char.
But I will add option to make check accept explicit cast.

**Comment by carlosgalvezp** [comment]
@firewave can you ellaborate on what you think is the problem? The cast does not solve the problem that the check is helping with, why should it not warn? Also, if you already have a cast, why not cast to the right type instead?

**Comment by firewave** [comment]
That cast does what it is supposed to do - enforce the usage of the ASCII value: https://godbolt.org/z/o746bxKMK

I chose that particular type since I did not want to cast from unsigned to signed.

**Comment by carlosgalvezp** [comment]
In that case the correct cast is to "char". There are 3 types of char, signed/unsigned char are numeric, and plain char is for text. Allowing such a cast would make sense as the intention of the author is clear.

**Comment by firewave** [comment]
> There are 3 types of char, signed/unsigned char are numeric, and plain char is for text.

What?

**Comment by carlosgalvezp** [comment]
> What?

https://eel.is/c++draft/basic#fundamental-7

> Type char is a distinct type that has an implementation-defined choice of “signed char” or “unsigned char” as its underlying type

Also:
https://stackoverflow.com/questions/16503373/why-is-char-distinct-from-signed-char-and-unsigned-char

`signed char` and `unsigned char` are _numeric_ types, i.e. intended to be used in arithmetic computations. Plain `char` is not a numeric type, but instead intended to represent a text character.

> I did not want to cast from unsigned to signed.

What problem do you get by doing so? It seems `std::basic_ostream` handles all 3 equally:
https://en.cppreference.com/w/cpp/io/basic_ostream/operator_ltlt2

**Comment by carlosgalvezp** [comment]
The docs also clearly state that in order to clearly express intent one should cast to `char`:

> Or cast to char to explicitly indicate that output should be a character.

**Comment by firewave** [comment]
Ugh, more homework...

> > I did not want to cast from unsigned to signed.
> 
> What problem do you get by doing so?

I did not want to unnecessarily cast to a different signedness (the codebase in questions is *really* bad with that and I did not want to add to the pile of stuff I need to review at some point in the future).

**Comment by HerrCai0907** [comment]
If you want an integer output here, then cast to uint32 is better. if you want a char, then cast to char is more suitable. actually there are no signed and unsigned in ascii level.

**Comment by carlosgalvezp** [comment]
> I did not want to unnecessarily cast to a different signedness

It appears the input is cast to `char` regardless, so it should be harmless for you to do so:

https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/ostream.h#L659
https://github.com/llvm/llvm-project/blob/main/libcxx/include/__ostream/basic_ostream.h#L385

**Comment by NagyDonat** [comment]
> > > I did not want to cast from unsigned to signed.
> > 
> > 
> > What problem do you get by doing so?
> 
> I did not want to unnecessarily cast to a different signedness (the codebase in questions is _really_ bad with that and I did not want to add to the pile of stuff I need to review at some point in the future).

In my opinion this is a valid concern -- even if `ostream` eventually casts everything to `char` (as an implementation detail), it's probably a good idea to avoid this unsigned -> signed cast in the codebase, because they are "red flags" in some sense and make it more difficult to understand the code and determine whether it's correct.

Also, IMHO it's very counter-intuitive that `std::cout << static_cast<char>(v);` works but `std::cout << static_cast<unsigned char>(v);` does not work. If you insist on preserving the current behavior, then this difference must be highlighted in the [documentation of the check](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unintended-char-ostream-output.html).

**Comment by carlosgalvezp** [comment]
> because they are "red flags" in some sense

I don't feel there's a strong, objective argumentation here, only that "casting to signed is bad". Why? What exactly is the problem? What does the Standard say? We must have answers to these questions if we want to enable this behavior.

> IMHO it's very counter-intuitive

It's not counter-intuitive if you consider that `unsigned char` is a *numeric type*, but `char` is a *character type*. Different types, different use cases.

If we want to allow printing an `unsigned char`, there's no way to distinguish that from an `std::uint8_t`. What would be the point of the check in that case? What is the blocker for putting a NOLINT suppression, and writing a good comment explaining the problematic code?

**Comment by NagyDonat** [comment]
> > because they are "red flags" in some sense
> 
> I don't feel there's a strong, objective argumentation here, only that "casting to signed is bad". Why? What exactly is the problem?

I do not say that "casting to signed is bad", I only say that it increases complexity. Casting between signed/unsigned types can introduce difficult-to-find bugs, because it is "hidden" (has no effect) when the value is positive and not too large, but can introduce corner cases where huge values or negative values suddenly appear and cause unexpected results. As I want to avoid these kinds of errors, I read `signed` ↔ `unsigned` casts with suspicion (especially if I'm trying to find the source of a bug), so I prefer code that doesn't use them superfluously.

> What does the Standard say?

`signed` ↔ `unsigned` conversions are completely well-defined, and there are many situations where they're the right tool for the job. I'm just saying that they are a bit complex, so I sympathize with using them sparingly, only when they're needed.

> > IMHO it's very counter-intuitive
> 
> It's not counter-intuitive if you consider that `unsigned char` is a _numeric type_, but `char` is a _character type_. Different types, different use cases.

My understanding is that `char`, `signed char` and `unsigned char` are all primarily used to represent _characters_ (as their name suggests) and it's preferable to say `int8_t` or `uint8_t` when I'm speaking about _numerical_ values. Can you cite a Standard section (or widespread style guide etc.) which says that `unsigned char` is numerical and not a character type?

> If we want to allow printing an `unsigned char`, there's no way to distinguish that from an `std::uint8_t`. What would be the point of the check in that case?

In my opinion the point of this check is that it highlight _ambiguous_ situations where the programmer didn't think about this question and will probably be surprised. If there is an explicit cast, then the check should accept it as the explicit decision of the programmer.

Also, it is possible to distinguish `unsigned char` from `std::uint8_t`, because type aliases are visible in the AST. (Conversely, if you don't look at the type alias name, then IIUC you cannot distinguish `char` – which is usually `signed char` on modern compilers – from `std::int8_t`.) In my opinion "do not report anything if an explicit cast states the programmer intent" is probably the best and most conservative approach, but if we want to report some of the expressions with casts, then checking the type alias name would be much cleaner instead of just checking "is this equivalent with `char`".

> What is the blocker for putting a NOLINT suppression, and writing a good comment explaining the problematic code?

There is no hard blocker, NOLINT is always an option. However I think it is very helpful if a `bugprone` check (which reports code that is easy to misunderstand, but can be correct) only reports cases where it appears that the programmer didn't think about the question – and suppresses reports where there is clear sign of programmer intent. (If the programmer thought about it and there _is_ a legitimate use, then it's nice to assume that the programmer was competent and decided to use it.)

**Comment by HerrCai0907** [comment]
So I think the best solution maybe not suppress warning for explicit cast. instead, we should provide option to suppress warning for the original variable type (with desurger) is `unsigned char` `signed char`.

**Comment by carlosgalvezp** [comment]
> Also, it is possible to distinguish unsigned char from std::uint8_t

Yes, this is a good point, of course the compiler knows everything :) 

Giving it some more thought, I believe I agree on making this categorization:

* `char`, `unsigned char` and `signed char` are "character types".
* `int8_t`, `uint8_t` are "numerical types".

I'm starting to think that maybe the check should warn only about "numeric" types, like `int8_t` and `uint8_t`, and not warn at all about `char` types. Or possibly have a `StrictMode` option that also warns about anything else than `char`.

I.e.:

```
cout << uint8_var;  // NOK
cout << int8_var; // NOK
cout << unsigned_char_var;  // OK (NOK if strict mode)
cout << signed_char_var;  // OK (NOK if strict mode)
cout << char_var; // OK, always
```

Casting uint8_t to unsigned char should make the check pass then, if not in strict mode. Or is there a use case for wanting the strict mode *and* the explicit cast?
 
WDYT?
